로그인 회원가입 PR 리뷰 요청드립니다.

yoonmin-kim opened the pull request:
## 설명
- 로그인, 회원가입 기능

## 작업사항
- 카카오/베이직 로그인 기능
- 카카오/베이직 회원가입 기능

## 리뷰포인트
- 엔티티
- 코드컨벤션
- 패키지구조

yoonmin-kim committed:
feat: 카카오톡 로그인 기능

yoonmin-kim committed:
feat: JWT 발급 및 인증필터 기능

내용:
- 카카오 인증이 완료되면 access, refresh 토큰발급
- 발급 된 access 토큰으로 인증처리
- refresh 토큰으로 access 토큰 갱신

yoonmin-kim committed:
feat: 로그인 및 회원가입 기능

내용:
- 카카오로 회원가입시 SignUpType.KAKAO
- 카카오로 회원가입시 BASIC 로그인 및 회원가입 불가능

yoonmin-kim committed:
feat: 이메일 인증기능 및 Instant 포맷 추가

내용:
- 이메일 인증시 해당 메일로 인증코드를 발송한다
- 레디스에 key:value=이메일:인증코드 형식으로 유효시간을 갖고 저장한다

yoonmin-kim committed:
feat: flyway 적용

yeoonjae commented on src/main/java/com/bside/threepick/config/RedisConfig.java:
```diff
@@ -0,0 +1,43 @@
+    return new LettuceConnectionFactory(host, port);
+  }
+
+  @Bean
+  public RedisTemplate<?, ?> redisTemplate() {
+    RedisTemplate<byte[], byte[]> redisTemplate = new RedisTemplate<>();
+    redisTemplate.setConnectionFactory(redisConnectionFactory());
+    redisTemplate.setKeySerializer(new StringRedisSerializer());
+    redisTemplate.setValueSerializer(new StringRedisSerializer());
+    return redisTemplate;
+  }
+
+  @Bean
+  public StringRedisTemplate stringRedisTemplate() {
+    StringRedisTemplate stringRedisTemplate = new StringRedisTemplate();
+    stringRedisTemplate.setKeySerializer(new StringRedisSerializer());
+    stringRedisTemplate.setValueSerializer(new StringRedisSerializer());
+    stringRedisTemplate.setConnectionFactory(redisConnectionFactory());
+    return stringRedisTemplate;
+  }
```
오 윤민님 redis 까지 모두 구현하셨군요! 👍

yeoonjae commented on src/main/java/com/bside/threepick/domain/account/controller/AccountController.java:
```diff
@@ -1,7 +1,56 @@
 import org.springframework.web.bind.annotation.RestController;
 
+@RequiredArgsConstructor
+@RequestMapping("/api/account")
```
account`s` 와 같이 복수형을 사용하는 것은 어떤가요? [rest API ](https://blog.pumpkin-raccoon.com/115) 관련 블로그를 봤는데 단수보다는 복수형을 많이 사용한다고하여 제안드립니다!

yeoonjae commented on src/main/java/com/bside/threepick/domain/account/controller/AccountController.java:
```diff
@@ -1,7 +1,56 @@
+    return ResponseEntity.ok(accountService.findAccountResponseByEmail(email));
+  }
+
+  @GetMapping("/{email}/auth")
```
이메일로 인증코드를 보내는 것이니 `@PostMapping` 어떠신가요?

yeoonjae commented on src/main/java/com/bside/threepick/domain/account/controller/AccountController.java:
```diff
@@ -1,7 +1,56 @@
+  @GetMapping("/{email}/{code}/auth")
+  public ResponseEntity<EmailAuthResponse> emailAuth(@PathVariable String email, @PathVariable String code) {
+    return ResponseEntity.ok(new EmailAuthResponse(accountService.isAuthenticatedEmail(email, code)));
+  }
```
{email}/check-code 와 같은 url 사용하고 code 는 파라미터로 받는 것이 더 좋을 것 같습니다!

yeoonjae commented on src/main/java/com/bside/threepick/domain/account/dto/SignUpRequest.java:
```diff
@@ -0,0 +1,32 @@
+  @Size(min = 8, max = 20, message = "영문, 숫자를 포함하여 8자~20자의 비밀번호를 입력해주세요.")
+  private String password;
+  @Size(min = 8, max = 20, message = "영문, 숫자를 포함하여 8자~20자의 비밀번호를 입력해주세요.")
+  private String passwordCheck;
```
비밀번호 체크 필드는 FE 단에서 확인할 것 같아서 별도로 존재하지 않아도 될 것 같다는 생각이 듭니당! 😁

yeoonjae commented on src/main/java/com/bside/threepick/domain/account/dto/SignUpRequest.java:
```diff
@@ -0,0 +1,32 @@
+
+  private boolean isNotSamePasswordAndPasswordCheck() {
+    return !password.equals(passwordCheck);
+  }
```
요 로직도 FE 단에서 체크할 로직으로 보여집니다 🙂

yeoonjae commented on src/main/java/com/bside/threepick/domain/account/reposiroty/AccountRepository.java:
```diff
@@ -0,0 +1,17 @@
+
+public interface AccountRepository extends JpaRepository<Account, Long> {
+
+  @Query("select a from Account a where a.email =:email")
```
JPQL은 지워도 될 것 같습니다 !

yoonmin-kim committed:
refactor: request-mapping 변경 및 Account에 컬럼 추가

yoonmin-kim commented:
연재님 PR확인 부탁드려요 :)

yoonmin-kim committed:
refactor: Account 컬럼타입 변경

yeoonjae commented on src/main/java/com/bside/threepick/domain/account/controller/AccountController.java:
```diff
@@ -1,7 +1,57 @@
+  @PostMapping
+  public ResponseEntity<AccountResponse> signup(@Validated @RequestBody SignUpRequest signUpRequest) {
+    AccountResponse accountResponse = accountService.signUp(signUpRequest);
+    return ResponseEntity.created(URI.create("/api/account/" + accountResponse.getEmail()))
```
`/api/accounts/`가 되어야 할 것 같습니다.

yeoonjae commented on src/main/java/com/bside/threepick/domain/account/controller/AccountController.java:
```diff
@@ -1,7 +1,57 @@
+  public ResponseEntity<AccountResponse> updateTimeValue(@Validated @RequestBody TimeValueRequest timeValueRequest) {
+    accountService.updateTimeValue(timeValueRequest);
+    return ResponseEntity.ok(accountService.findAccountResponseByEmail(timeValueRequest.getEmail()));
+  }
```
이 API 는 timeValue 를 update 하는 것으로 보여지는데 `accountService.findAccountResponseByEmail` 을 호출하는 이유가 있을까요?

yoonmin-kim replied:
사용자 정보가 필요할때는 다시 요청을 주는게 맞다고 판단하여 남겨주신 의견과 같이 수정했습니다 👍

yeoonjae commented on src/main/resources/db/migration/V1__Create Table Account.sql:
```diff
@@ -0,0 +1,19 @@
+create table account (
+     id bigint not null auto_increment,
+     created_date datetime(6),
+     modified_date datetime(6),
+     email varchar(255) not null,
+     nick_name varchar(255),
+     password varchar(255),
+     signup_type varchar(255) not null,
+     status varchar(255) default 'ACTIVE' not null,
+     last_login_date datetime(6),
+     time_value integer,
+     total_value integer,
+     next_time_value integer default '0',
+     next_time_value_date datetime(6),
+     change_count integer default '0',
+     primary key (id)
+ ) engine=InnoDB;
+
+alter table account add constraint UK_ACCOUNT_EMAIL unique (email);
```
테이블 설계가 노션에 정의되어있는 것과 다른 것 같아요 확인 부탁드립니다.  🙇‍♀️

yeoonjae commented on src/main/resources/db/migration/V1__Create Table Account.sql:
```diff
@@ -0,0 +1,19 @@
+create table account (
+     id bigint not null auto_increment,
+     created_date datetime(6),
+     modified_date datetime(6),
```
```sql
`created_date` datetime     NOT NULL DEFAULT current_timestamp(),
`modified_date` datetime     NOT NULL DEFAULT current_timestamp(),
```
로 변경하는 게 좋을 것 같습니다!

yeoonjae commented on src/main/java/com/bside/threepick/domain/account/controller/AccountController.java:
```diff
@@ -1,7 +1,57 @@
+
+  private final AccountService accountService;
+
+  @GetMapping("/{email}")
```
email 을 pathvariable 로 넘겨받을 때 [이런 이슈](https://soheemon.tistory.com/entry/Spring-RestController에서-PathVariable로-email을-넘길때-이슈)가 있다고 하네요..!! 다른 좋은 방법이 있는지 한 번 생각해봐도 좋을 것 같습니다!

yoonmin-kim replied:
이부분은 postman이랑 브라우저로 테스트 해봤는데 문제가 없는거 같아요. 혹시라도 문제가 확인되면 링크주신 내용과 같이 {email:.+} 를 사용하도록 수정하는게 어떨까요? 간단하게 해결이 가능한거 같아서 다행이네요!

yoonmin-kim committed:
feat: 로그인시 lastLoginDate 를 update

내용:
- 테이블설계 변경
- 로그인시 lastLoginDate 를 update 한다
- sendRedirect 시 토큰을 쿼리스트링으로 넘긴다

yoonmin-kim commented:
연재님 꼼꼼한 리뷰 감사합니다! 다시한번 확인 부탁드려요 :)

yeoonjae commented on src/main/java/com/bside/threepick/domain/security/filter/JwtAuthFilter.java:
```diff
@@ -0,0 +1,47 @@
+      String role = tokenService.getRole(token);
+
+      UserDetails userDetails = User.builder()
+          .username(email)
+          .password("")
+          .authorities(role)
+          .build();
```
윤민님 혹시, accountId 도 JwtToken 에 같이 넣어주실 수 있나요?

yoonmin-kim committed:
feat: jwt accountId 추가

내용:
- jwt accountId 추가
- 한시간의 가치 변경시 changeCount 증가한다
- Instant -> LocalDateTime 변경

yoonmin-kim commented:
연재님 accountId jwt 에 포함하도록 수정되었습니다. 확인 부탁드려요 :D

yeoonjae reviewed (approved):
accountId 도 token 값에 같이 담길 수 있도록 수정하신 부분 확인했습니다. 수고하셨습니다!
